### PR TITLE
feat(ga): add a new datasource to get the list of the listeners

### PR DESCRIPTION
**Pull request by ruwenqiang123**
**What this PR does / why we need it**:

Add a new datasource to get the list of the listeners.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. support a new datasource to get the list of the listeners.
2. support related document and acceptance test.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
root@ecs-ruwenqiang:/home/huawei/go/src/github.com/huaweicloud/terraform-provider-huaweicloud (dev_listener)$ make testacc TEST="./huaweicloud/services/acceptance/ga" TESTARGS="-run TestAccDatasourceListeners_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/ga -v -run TestAccDatasourceListeners_basic -timeout 360m -parallel 4
=== RUN   TestAccDatasourceListeners_basic
=== PAUSE TestAccDatasourceListeners_basic
=== CONT  TestAccDatasourceListeners_basic
--- PASS: TestAccDatasourceListeners_basic (105.80s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/ga        105.853s
```

### CODE REVIEW

File: huaweicloud/services/acceptance/ga/data_source_huaweicloud_ga_listeners_test.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+	)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:          func() { acceptance.TestAccPreCheck(t) },
+		ProviderFactories: acceptance.TestAccProviderFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceListeners_basic(name),
+				Check: resource.ComposeTestCheckFunc(
+					dc.CheckResourceExists(),
```
Comment (deer-hang) [inline huaweicloud/services/acceptance/ga/data_source_huaweicloud_ga_listeners_test.go:41]: Please check the Computed field values.

File: huaweicloud/services/ga/data_source_huaweicloud_ga_listeners.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+func filterListListenersResponseBody(all []interface{}, d *schema.ResourceData) []interface{} {
+	rst := make([]interface{}, 0, len(all))
+	for _, v := range all {
+		if param, ok := d.GetOk("protocol"); ok &&
+			fmt.Sprint(param) != utils.PathSearch("protocol", v, nil) {
+			continue
+		}
+
+		portRanges := utils.PathSearch("port_ranges", v, nil).([]map[string]interface{})
```
Comment (deer-hang) [inline huaweicloud/services/ga/data_source_huaweicloud_ga_listeners.go]: This line may cause panic error.

File: huaweicloud/services/ga/data_source_huaweicloud_ga_listeners.go [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+		if param, ok := d.GetOk("to_port"); ok && len(filterByPortRange(portRanges, param.(int), "to_port")) == 0 {
+			continue
+		}
+
+		rst = append(rst, v)
+	}
+	return rst
+}
+
+func filterByPortRange(portRanges []map[string]interface{}, target int, port string) []map[string]interface{} {
```
Comment (deer-hang) [inline huaweicloud/services/ga/data_source_huaweicloud_ga_listeners.go]: I think the return value of this method only needs to be a boolean value.

File: huaweicloud/services/acceptance/ga/data_source_huaweicloud_ga_listeners_test.go (hunk 1 above)
Comment (ruwenqiang123) [reply huaweicloud/services/acceptance/ga/data_source_huaweicloud_ga_listeners_test.go:41]: done

File: huaweicloud/services/ga/data_source_huaweicloud_ga_listeners.go (hunk 2 above)
Comment (ruwenqiang123) [reply huaweicloud/services/ga/data_source_huaweicloud_ga_listeners.go]: done

File: huaweicloud/services/ga/data_source_huaweicloud_ga_listeners.go (hunk 3 above)
Comment (ruwenqiang123) [reply huaweicloud/services/ga/data_source_huaweicloud_ga_listeners.go]: done

File: huaweicloud/services/ga/data_source_huaweicloud_ga_listeners.go [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+func filterByListenPort(portRanges []map[string]interface{}, target int, port string) bool {
+	rst := make([]map[string]interface{}, 0, len(portRanges))
+	for _, v := range portRanges {
+		if int(utils.PathSearch(port, v, 0).(float64)) == target {
+			rst = append(rst, v)
+		}
+	}
+
+	return len(rst) == 0
+}
```
Comment (deer-hang) [inline huaweicloud/services/ga/data_source_huaweicloud_ga_listeners.go]:
```suggestion
func filterByListenPort(portRanges []map[string]interface{}, target int, port string) bool {
	for _, v := range portRanges {
		if int(utils.PathSearch(port, v, float64(0)).(float64)) == target {
			return true
		}
	}
	return false
}
```
I think filter with parameters `from_port` and `to_port` is unnecessary. Can these two filter conditions be deleted?

Comment (ruwenqiang123) [reply huaweicloud/services/ga/data_source_huaweicloud_ga_listeners.go]: done

**Comment by ruwenqiang123** [comment]
```
root@ecs-ruwenqiang:/home/huawei/go/src/github.com/huaweicloud/terraform-provider-huaweicloud (dev_listener)$ make testacc TEST="./huaweicloud/services/acceptance/ga" TESTARGS="-run TestAccDatasourceListeners_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/ga -v -run TestAccDatasourceListeners_basic -timeout 360m -parallel 4
=== RUN   TestAccDatasourceListeners_basic
=== PAUSE TestAccDatasourceListeners_basic
=== CONT  TestAccDatasourceListeners_basic
--- PASS: TestAccDatasourceListeners_basic (95.11s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/ga        95.152s
```

**Pull request merged**